res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,621 @@
+    midi.sendShortMsg(0x94 + offset, 0x0D, Prime4.rgbCode.whiteDark);
+    //midi.sendShortMsg(0x94 + offset, 0x0E, Prime4.rgbCode.whiteDark); // Slicer mode, not implemented yet
+};
+
+Prime4.padSection.prototype = Object.create(components.ComponentContainer.prototype);
+
+Prime4.padSection.prototype.controlToPadMode = function(control) {
+    let mode;
+    switch (control) {
+    case Prime4.padMode.HOTCUE:
+        mode = this.modes.hotcue;
+        break;
+    case Prime4.padMode.LOOP:
+        if (this.currentMode === this.modes.loop) {
+            mode = this.modes.autoloop;
+        } else {
+            mode = this.modes.loop;
+        }
+        break;
+    case Prime4.padMode.ROLL:
+        mode = this.modes.roll;
+        break;
+    case Prime4.padMode.SLICER:
+        mode = this.modes.sampler;
+        break;
+    }
+
+    return mode;
+};
+
+Prime4.padSection.prototype.padModeButtonPressed = function(channel, control, value, _status, _group) {
+    if (value) {
+        this.setPadMode(control);
+    }
+};
+
+// Worry about parameter buttons later
+//Prime4.padSection.prototype.paramButtonPressed = function(channel, control, value, status, group) {};
+
+Prime4.padSection.prototype.setPadMode = function(control) {
+    const newMode = this.controlToPadMode(control);
+
+    // Exit early if requested mode is already active or unavailable
+    if (newMode === this.currentMode || newMode === undefined) {
+        return;
+    }
+
+    // Disable LED of current mode button
+    if (this.currentMode) {
+        midi.sendShortMsg(0x94 + this.offset, this.currentMode.ledControl, Prime4.rgbCode.whiteDark);
+
+        // Disconnect pads from current mode
+        this.currentMode.forEachComponent(function(component) {
+            component.disconnect();
+        });
+    }
+
+    // Enable LED of new mode to switch to
+    midi.sendShortMsg(0x94 + this.offset, newMode.ledControl, newMode.colour);
+
+    // Connect pads to new mode
+    newMode.forEachComponent(function(component) {
+        component.connect();
+        component.trigger();
+    });
+
+    this.currentMode = newMode;
+
+};
+
+Prime4.padSection.prototype.performancePad = function(channel, control, value, status, group) {
+    const i = (control - 0x0E);
+    this.currentMode.pads[i].input(channel, control, value, status, group);
+};
+
+Prime4.hotcueMode = function(deck, offset) {
+    components.ComponentContainer.call(this);
+    this.ledControl = Prime4.padMode.HOTCUE;
+    this.colour = Prime4.rgbCode.blue;
+    this.pads = new components.ComponentContainer();
+    for (let i = 1; i <= 8; i++) {
+        this.pads[i] = new components.HotcueButton({
+            number: i,
+            group: deck.currentDeck,
+            midi: [0x94 + offset, 0x0E + i],
+            colorMapper: Prime4ColorMapper,
+            on: this.colour,
+            off: Prime4.rgbCode.whiteDark,
```

Comment:
I saw this in the DJ505 but wasn't sure what it did, so I left it out for now. What does `outConnect` actually do?

Reply:
It's loosely described in the componentJS docs: https://github.com/mixxxdj/mixxx/wiki/Components-JS#component

let me summarize it like this: You have an `inConnect` that configures whether the CO controlled by the component should receive midi input. In reality, there is no `inConnect` and that binding is actually done via the XML. _But we do have an `outConnect` whose job is to specify that same relationship in the other direction, so whether the value of a CO should control the specified LED_. By default, `outConnect` is `true` since you actually want that, but in the case of the padmodes, where multiple components control the same LED, you need to be careful that only a single component is ever actually connected. Otherwise the different components would fight over the same LED. For that reason, in the padmodes, we have them disconnected by default and only connect a single mode explicitly after all the modes have been instantiated. Does that make sense?

Reply:
Ahhh right, yes that makes sense. Thanks for clarifying!